rust/tw_proto/src/ffi.rs

```diff
@@ -0,0 +1,104 @@
+#![allow(clippy::missing_safety_doc)]
+
+use std::borrow::Cow;
+use std::ffi::c_char;
+use tw_memory::c_string_standalone;
+use tw_memory::ffi::c_byte_array::{CByteArray, CByteArrayResult};
+use tw_memory::ffi::c_byte_array_ref::CByteArrayRef;
+use tw_memory::ffi::c_result::ErrorCode;
+
+#[repr(C)]
+pub enum CProtoError {
+    Ok = 0,
+    Internal = 1,
+    ErrorDeserializingMsg = 2,
+    ErrorSerializingMsg = 3,
+}
+
+impl From<CProtoError> for ErrorCode {
+    fn from(e: CProtoError) -> Self {
+        e as ErrorCode
+    }
+}
+
+/// Takes a serialized `Ethereum::Proto::SigningInput` message, deserializes it and returns a serialized message back.
+/// This FFI is used within integration tests.
+/// \param input *non-null* byte array.
+/// \param input_len length of the input byte array.
+/// \return C-compatible result with a C-compatible byte array.
+#[no_mangle]
+pub unsafe extern "C" fn pass_eth_signing_msg_through(
+    input: *const u8,
+    input_len: usize,
+) -> CByteArrayResult {
+    let data = CByteArrayRef::new(input, input_len)
+        .to_vec()
+        .unwrap_or_default();
+
+    let msg: crate::Ethereum::Proto::SigningInput = match crate::deserialize(&data) {
```

Comment:
Is it possible to use generic here? otherwise we have to create boiler code for each proto SigningInput?

Reply:
These FFIs are needed just to make sure if some of the Protobuf messages are serialized and deserialized correctly between C++ and Rust implementations.
We don't need to add new FFIs in this module anymore. Moreover, when we integrate one or two blockchains, the serialization correctness will be implicit checked, and we'll be able to remove these FFIs.